File: workdir/src/TestDriver.kt
// -*- coding: utf-8; -*-

package com.k650250.odb.testing

import com.k650250.odb.ODBFile
import java.sql.DriverManager
import java.sql.ResultSet
import java.sql.SQLException
import java.util.Scanner

fun main() {
    // ロギングの環境設定ファイル (※ HSQLDB-1.8.0.10 では使用しない)
    //System.setProperty("java.util.logging.config.file", "logging.properties")

    try {
        Class.forName("org.hsqldb.jdbcDriver")

        ODBFile.open("sample.odb").use { odbFile ->
            val con = DriverManager.getConnection(odbFile.toUrl(), "sa", "")
            val st = con.createStatement();
            var sql: String
            var records: ResultSet

            println("[更新前のデータ一覧]")
            //sql = """SELECT * FROM "t_sample" """  // テーブル
            //sql = """SELECT * FROM "v_sample" """  // ビュー
            //sql = odbFile.getQuery("q_sample").toString()         // クエリー (問合せ文中にクエリー名が含まれない場合)
            sql = odbFile.getQuery("q_SAMPLE").expand().toString()  // クエリー (問合せ文中にクエリー名が含まれる場合)
            records = st.executeQuery(sql)
            while (records.next()) {
                println("${records.getString("key")}\t${records.getString("value")}")
            }
            records.close()

            // データ挿入
            sql = """INSERT INTO "t_sample"("value") VALUES(?)"""
            val prep = con.prepareStatement(sql)
            val sc = Scanner(System.`in`, System.getProperty("native.encoding"))
            print("追加データを入力してください: ")
            prep.setString(1, sc.nextLine())
            prep.executeUpdate()
            sc.close()
            prep.close()

            println("[更新後のデータ一覧]")
            sql = """SELECT * FROM "t_sample" """
            records = st.executeQuery(sql)
            while (records.next()) {
                println("${records.getString("key")}\t${records.getString("value")}")
            }
            records.close()

            st.close()
            con.close()
        }
    } catch (e: ClassNotFoundException) {
        e.printStackTrace()
        System.err.println("org.hsqldb.jdbcDriver が見つかりません。")
    } catch (e: SQLException) {
        e.printStackTrace()
        System.err.println("考えられる原因:")
        System.err.println("- hsqldb.jar が正常に読み込まれていない")
        System.err.println("- このデータベースを別のプロセスが使用中")
        System.err.println("- 問合せ文中にクエリー名が含まれる")
    } catch (e: Exception) {
        e.printStackTrace()
    }
}

File: workdir/src/TestDriver.rb
# -*- coding: utf-8; -*-

require 'java'
java_import 'com.k650250.odb.ODBFile'
java_import java.lang.ClassNotFoundException
java_import java.lang.System
java_import java.sql.DriverManager
java_import java.sql.SQLException
java_import java.util.Scanner

# ロギングの環境設定ファイル (※ HSQLDB-1.8.0.10 では使用しない)
#System.setProperty("java.util.logging.config.file", "logging.properties")

def closing(thing)
    if block_given?
        yield thing
    end
ensure
    thing.close
end

begin
    loader = java.lang.Thread.current_thread.getContextClassLoader
    java.lang.Class.forName("org.hsqldb.jdbcDriver", true, loader)

    closing(ODBFile.open("sample.odb")) do |odbFile|
        closing(DriverManager.getConnection(odbFile.toUrl, "sa", "")) do |con|
            closing(con.createStatement()) do |st|
                puts "[更新前のデータ一覧]"
                #sql = 'SELECT * FROM "t_sample" '  # テーブル
                #sql = 'SELECT * FROM "v_sample" '  # ビュー
                #sql = odbFile.getQuery("q_sample").toString         # クエリー (問合せ文中にクエリー名が含まれない場合)
                sql = odbFile.getQuery("q_SAMPLE").expand.toString  # クエリー (問合せ文中にクエリー名が含まれる場合)
                closing(st.executeQuery(sql)) do |records|
                    while records.next do
                        puts "#{records.getString("key")}\t#{records.getString("value")}"
                    end
                end

                # データ挿入
                sql = 'INSERT INTO "t_sample"("value") VALUES(?)'
                closing(con.prepareStatement(sql)) do |prep|
                    closing(Scanner.new(System.in, System.getProperty("native.encoding"))) do |sc|
                        puts "追加データを入力してください。"
                        prep.setString(1, sc.nextLine)
                        prep.executeUpdate
                    end
                end

                puts "[更新後のデータ一覧]"
                sql = 'SELECT * FROM "t_sample" '
                closing(st.executeQuery(sql)) do |records|
                    while records.next do
                        puts "#{records.getString("key")}\t#{records.getString("value")}"
                    end
                end
            end
        end
    end
rescue ClassNotFoundException => e
    e.printStackTrace
    STDERR.puts "考えられる原因:"
    STDERR.puts "- hsqldb.jar がクラス・パス指定されていない"
    STDERR.puts "- odb.jar がクラス・パス指定されていない"
rescue SQLException => e
    e.printStackTrace
    STDERR.puts "考えられる原因:"
    STDERR.puts "- hsqldb.jar が正常に読み込まれていない"
    STDERR.puts "- このデータベースを別のプロセスが使用中"
    STDERR.puts "- 問合せ文中にクエリー名が含まれる"
end
